deepchecks/nlp/metric_utils/token_classification.py

```diff
@@ -0,0 +1,198 @@
+                      ) -> t.Tuple[t.List[t.List[str]], t.List[t.List[str]]]:
+        """Fit (if necessary) and transform the labels and predictions of a specific dataset."""
+        if self.processed_labels is None or self.processed_predictions is None:
+            self.fit(y_true, y_pred)
+        return self.processed_labels, self.processed_predictions
+
+    def _spans_to_token_list(self, y_true: TTokenLabel, y_pred: 'TTokenPred',
+                             ) -> t.Tuple[t.List[t.List[str]], t.List[t.List[str]]]:
+        """Convert from predictions and labels in span format to a list of token annotations.
+
+        Parameters:
+        -----------
+            y_true: TTokenLabel
+                Labels as provided by the user in the accepted format
+            y_pred: TTokenPred
+                Predictions as provided by the user in the accepted format
+
+        Returns:
+        --------
+            A tuple of processed labels and processed predictions.
+        """
+        if not len(y_true) == len(y_pred):
+            raise DeepchecksValueError(f'Must have both a label entry and a prediction entry for each sample, but found '
+                                       f'{len(y_true)} label entries and {len(y_pred)} prediction entries')
+
+        labels = []
+        predictions = []
+        for sample_idx in range(len(y_true)):
+            sample_y_true = sorted(y_true[sample_idx], key=lambda x: x[1])
+            sample_y_pred = sorted(y_pred[sample_idx], key=lambda x: x[1])
+            sample_labels = [annotation[0] for annotation in sample_y_true]
+            labels.append(sample_labels)
+
+            label_spans = [(annotation[1], annotation[2]) for annotation in sample_y_true]
+            pred_spans = [(annotation[1], annotation[2]) for annotation in sample_y_pred]
+
+            if len(sample_y_true) > 0:
+                # Indices of where to place elements of pred_spans in label_spans so that it's sorted
+                sorted_index = np.searchsorted(list(map(lambda x: x[0], pred_spans)),
+                                               list(map(lambda x: x[0], label_spans)), side='left')
+                # Elements of sample_y_pred that match their counterparts in label_spans
+                pred_index = np.take(np.arange(len(pred_spans)), sorted_index, mode='clip')
+                bool_matching = np.array(list(map(hash, pred_spans)))[pred_index] == np.array(list(map(hash, label_spans)))
+                # Filter out predictions that do not match their corresponding labels or their threshold is too low.
+                # Predictions that where filtered are filled with 'O'.
+                sample_predictions = \
+                    [sample_y_pred[pred_index[idx]][0] if
+                     ((sample_y_pred[pred_index[idx]][3] >= self.proba_threshold) and bool_matching[idx]) else 'O'
+                     for idx in range(len(sample_y_true))]
```

Comment:
This is hard to understand, so I cant properly review.
easy ideas to improve:
Better variable names, explain at start what is the expected format to be in sample_prediction